### PR TITLE
first pass at Gramian training for OLS

### PR DESCRIPTION
**Pull request by adienes**
Ref: https://github.com/JuliaAI/MLJLinearModels.jl/issues/145

This is a very basic initial implementation. I only checked the simplest possible case which is `Lasso` with a single target, no intercept, etc. etc. but the following will work as expected:

```
# X, y some data
lasso = LassoRegression(0.01, fit_intercept=false)
fit_gram(lasso, X'X, X'y; n=100)
```

Please suggest any cleanups that need to be made to make this robust addition to package. I know it is a little quickly done but I wasn't sure what to focus on to clean

Need to still
1. decide what the right signature is to access training on XX, Xy
2. either make inputs more strict (with loud `NotImplementedErrors`) or make implementation more generic, right now the functions will accept more than they can actually handle
3. add test cases

### CODE REVIEW

**Comment by adienes** [comment]
@tlienart thank you for the comments and apologies for the delay

I made the call signature like this
```
θ_gram = fit(enr; data=(; XX, Xy, n), solver=FISTA(gram=true))
```
I did not want to pun the arguments of `X` and `y` so I added a kwarg `data`. the reasons are
1. when using from MLJ it would be surprising if accidentally used cross validation / fold splitting on `XX` and `Xy`
2. it might be desirable to support `gram=true` even when passing in `X` and `y` and then the kernels will be precomputed by the library. this will also allow for fit_intercept which I left unimplemented here

**Comment by adienes** [comment]
for `X` of size `(100000, 200)` , the performance is about `200x` better for both runtime and memory usage
```
julia> @benchmark lin.fit(enet, X, y; data, solver=FISTA(gram=true, max_iter=5000))
BenchmarkTools.Trial: 257 samples with 1 evaluation.
 Range (min … max):  19.109 ms …  21.066 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     19.522 ms               ┊ GC (median):    1.19%
 Time  (mean ± σ):   19.524 ms ± 254.764 μs  ┊ GC (mean ± σ):  0.77% ± 0.60%

        ▂▂ ▂▃ ▁  ▂   ▆▅ ▆ ▅█▁▆▃▃▂▃▂                             
  ▄▆▄▃▄▄██▇██▃█▄▆█▇▅███▅█▅██████████▅▄▇▃▅▁▃▁▃▁▃▁▁▃▃▃▁▅▅▄▃▅▁▁▁▃ ▄
  19.1 ms         Histogram: frequency by time         20.1 ms <

 Memory estimate: 27.34 MiB, allocs estimate: 44058.

julia> @benchmark lin.fit(enet, X, y; data, solver=FISTA(gram=false, max_iter=5000))
BenchmarkTools.Trial: 1 sample with 1 evaluation.
 Single result which took 7.386 s (1.65% GC) to evaluate,
 with a memory estimate of 4.70 GiB, over 50418 allocations.
```

**Comment by adienes** [comment]
some remaining question marks

- what should happen if the user passes in `X` and `y` that do not actually correspond to `XX` and `Xy` given in `data` ?
- the `objective` function does not have an appropriate signature here. should we add one?

**Comment by adienes** [comment]
was wondering if we could dust this off. is the API ok? what are the main concerns

File: src/utils.jl [hunk 1]
```diff
@@ -9,6 +9,29 @@ function check_nrows(X::AbstractMatrix, y::AbstractVecOrMat)::Nothing
...
+    !iszero(c) && throw(ArgumentError("Categorical loss not supported with Gramian kernel"))
+    glr.fit_intercept && throw(ArgumentError("Intercept not supported with Gramian kernel"))
+
+    if any(.!iszero.((X, y)))
+        all((
+            isapprox(X'X, data.XX; rtol=1e-5),
+            isapprox(X'y, data.Xy; rtol=1e-5),
+            length(y) == data.n
+        )) || throw(ArgumentError("Inputs `X` and `y` do not match inputs `XX` and `Xy`."))
+    end
```
Comment (tlienart) [inline src/utils.jl:30]: I might be misunderstanding this check, if a user provides `data` then is it the case that `XX` and `Xy` are (re)computed to check that `data` is correct?

**Review by tlienart — COMMENTED** [review]
Sorry I let this slip... I have only one question in `verify*` which is likely due to me not following the logic well there. Otherwise this seems great, thanks for your work!

File: src/utils.jl (hunk 1 above)
Comment (adienes) [reply src/utils.jl:30]: yes, that was the intention. if `data` only is provided, `XX, Xy` used as given. if `X, y` only provided, then `XX, Xy` recomputed and used. if both `data` and `X, y` provided, then `XX, Xy` are recomputed to catch the user from making silly errors, but the original values in `data` are used

note the the added method `fit(glr::GLR; kwargs...) = fit(glr, zeros((0,0)), zeros((0,)); kwargs...)` to allow calling without `X,y` at all. I changed the `iszero` check to `isempty` to make it clear those are dummy variables

**Comment by adienes** [comment]
hang on, tried to rebase on dev, something went funky

File: src/utils.jl (hunk 1 above)
Comment (tlienart) [reply src/utils.jl:30]: > if both data and X, y provided, then XX, Xy are recomputed to catch the user from making silly errors, but the original values in data are used

ok this is what I hadn't understood, thanks for the clarification

**Comment by tlienart** [comment]
will wait for CI to pass then merge & do a minor release with it, thanks again

**Pull request merged**